zproject/backends.py

```diff
@@ -624,6 +626,24 @@ def __init__(self) -> None:
         if settings.DEVELOPMENT and settings.FAKE_LDAP_MODE:  # nocoverage
             init_fakeldap()
 
+        if settings.LDAP_SYNC_GROUPS:
+            self._allZulipGroupNameMap = {group.name: group for group in UserGroup.objects.filter(realm_id = settings.LDAP_GROUP_SYNC_REALM_ID)}
+            # self._allLDAPGroupNames = {} # TODO if setting not set, get all ldap groups
+            self._allLDAPGroupNames = settings.LDAP_GROUP_SYNC_LIST
+
+            realm = Realm.objects.get(id=settings.LDAP_GROUP_SYNC_REALM_ID)
+
+            for group in self._allLDAPGroupNames:
+                if not UserGroup.objects.filter(name=group).exists():
+                    ldap_logger.debug(
+                        "creating zulip group %s from ldap for realm %s", group, realm.name
+                    )
+                    UserGroup.objects.create(
+                        name=group,
+                        description="LDAP import of " + group,
+                        realm=realm,
+                    )
```

Comment:
Why do we do this specific part in `__init__` instead of in `sync_groups_from_ldap`

Reply:
Also, I'm not sure if we really want to be creating a USerGroup just because it's in `LDAP_GROUP_SYNC_LIST` - maybe we should create on-demand when actually syncing a user that should belong in the group?

Reply:
> Why do we do this specific part in `__init__` instead of in `sync_groups_from_ldap`

`sync_groups_from_ldap` from ldap is invoked for each user, this only needs to be done once per run.

> Also, I'm not sure if we really want to be creating a USerGroup just because it's in `LDAP_GROUP_SYNC_LIST` - maybe we should create on-demand when actually syncing a user that should belong in the group?

Same issue as above, the logic for creating a missing group would need to be invoked for every user.

Reply:
I started implementing the on-demand way and it would add quite a bit of boilerplate and spaghetti code, this is what I would need to do:

```python

    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
        if not settings.LDAP_SYNC_GROUPS:
            return

        if user_profile.realm != settings.LDAP_GROUP_SYNC_REALM_ID:
            return

        try:
            ldap_logger.debug("found groups for user: %s", str(user_profile))
            ldap_group_name_set = set(ldap_user.group_names)

            member_group_names = [
                membership.user_group.name
                for membership in UserGroupMembership.objects.filter(user_profile=user_profile)
            ]

            ldap_logger.debug("user groups: %s", repr(member_group_names))

            for group_name in ldap_group_name_set:
                if group_name not in set(settings.LDAP_GROUP_SYNC_LIST):
                    continue # this groups shall not be synced!

                ldap_logger.debug("group: %s", str(group))

                group = UserGroup.objects.filter(name=group_name)

                if group.exists():
                    group = group.get()
                else:
                    ldap_logger.debug(
                        "creating zulip group %s from ldap for realm %s", group, user_profile.realm.name
                    )
                    group = UserGroup.objects.create(
                        name=group,
                        description="LDAP import of " + group,
                        realm=user_profile.realm,
                    )

                # we continue above if group should not be synced and create group if not exists no need to check the group again
                if not is_user_in_group(user_group=group, user=user_profile, direct_member_only=True):
                    ldap_logger.debug("add %s to %s", user_profile.full_name, str(group))
                    UserGroupMembership.objects.create(
                        user_group=group,
                        user_profile=user_profile,
                    )
                    do_send_user_group_members_update_event(event_name="add_members", user_group=group, user_profile_ids=[user_profile.id])

            for group in settings.LDAP_GROUP_SYNC_LIST:
                ##### TODO: I would need to check group existing here AGIN! <-----
                if is_user_in_group(user_group=group, user=user_profile, direct_member_only=True):
                    continue  # we remove here, so skip if in group

                if group in member_group_names:
                    ldap_logger.debug("remove %s from %s", user_profile.full_name, group)
                    UserGroupMembership.objects.get(
                        user_profile=user_profile,
                        user_group=self._all_zulip_group_name_map[group],
                    ).delete()

        except Exception as e:
            raise ZulipLDAPException(str(e)) from e

```

I think creating all groups at the beginning is the best way to go from both performance and code readability.